oauth2-client/src/main/java/org/springframework/security/oauth2/client/authentication/AuthorizationGrantTokenExchanger.java

```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright 2012-2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.oauth2.client.authentication;
+
+
+import org.springframework.security.oauth2.core.OAuth2AuthenticationException;
+import org.springframework.security.oauth2.core.protocol.TokenResponseAttributes;
+
+/**
+ * @author Joe Grandja
+ */
+public interface AuthorizationGrantTokenExchanger<T extends AuthorizationGrantAuthenticationToken>  {
```

Comment:
There will be different implementations of `AuthorizationGrantTokenExchanger` for each of the grant types `authorization_code`, `client_credentials` and `resource_owner_password` grant types.

For example, the `client_credentials` grant type would be `ClientCredentialsTokenExchanger` whoch would be associated with it's specific type of `AuthorizationGrantAuthenticationToken`, for example, `ClientCredentialsAuthenticationToken`